pkg/agent/nodeportlocal/rules/iptable_rule.go

```diff
@@ -46,37 +47,57 @@ func NewIPTableRules() *iptablesRules {
 
 // Init initializes IPTABLES rules for NPL. Currently it deletes existing rules to ensure that no stale entries are present.
 func (ipt *iptablesRules) Init() error {
-	return ipt.CreateChains()
+	if err := ipt.InitRules(); err != nil {
+		return fmt.Errorf("initialization of NPL iptables rules failed: %v", err)
+	}
+	return nil
 }
 
-// CreateChains creates the chain NodePortLocalChain in NAT table.
-// All DNAT rules for NPL would be added in this chain.
-func (ipt *iptablesRules) CreateChains() error {
-	err := ipt.table.EnsureChain(iptables.NATTable, NodePortLocalChain)
-	if err != nil {
-		return fmt.Errorf("IPTABLES chain creation in NAT table failed for NPL with error: %v", err)
+// InitRules creates the NPL chain and link them to the PREROUTING (for incoming
+// traffic) and OUTPUT chain (for locally-generated traffic). All NPL DNAT rules
+// will be added to this chain.
+func (ipt *iptablesRules) InitRules() error {
+	if err := ipt.table.EnsureChain(iptables.NATTable, NodePortLocalChain); err != nil {
+		return err
 	}
 	ruleSpec := []string{
-		"-p", "tcp", "-j", NodePortLocalChain,
+		"-p", "tcp", "-m", "addrtype", "--dst-type", "LOCAL", "-j", NodePortLocalChain,
 	}
-	err = ipt.table.EnsureRule(iptables.NATTable, iptables.PreRoutingChain, ruleSpec)
-	if err != nil {
-		return fmt.Errorf("IPTABLES rule creation in NAT table failed for NPL with error: %v", err)
+	if err := ipt.table.EnsureRule(iptables.NATTable, iptables.PreRoutingChain, ruleSpec); err != nil {
+		return err
+	}
+	if err := ipt.table.EnsureRule(iptables.NATTable, iptables.OutputChain, ruleSpec); err != nil {
+		return err
 	}
 	return nil
 }
 
+func buildRulesForPod(port int, podIP string) [][]string {
+	var rules [][]string
+	// For local traffic we need to masquerade the traffic to ensure that
+	// the gateway's IP is used as the source. Otherwise, traffic will be
+	// dropped by the SpoofGuard table in the OVS pipeline.
```

Comment:
Does the spoofguard table check source IP for traffic coming from antrea-gw0? I thought no as nodeport service with `externalTrafficPolicy=Local` will preserve client IP?

Reply:
Sorry the comment is a bit inaccurate. The SpoofGuard table may drop the ARP requests: on my Node without SNAT the ARP requests are sent with the IP address of the default network interface as the advertised address, and these packets are dropped by SpoofGuard. I was surprised by this behavior, until I found out about the `arp_announce` sysctl and its default value on Linux.

I then checked the kube-proxy behavior for locally initiated traffic (from the host) for a NodePort Service with `externalTrafficPolicy` set to `Local`, and it applies the masquerade mark.

Reply:
Thanks for the explanation. I'm suprised by the behavior too, thought ARP is independent to IP packet that triggers it. Good to learn this.